hail/src/main/scala/is/hail/expr/ir/RefEquality.scala

```diff
@@ -31,6 +31,9 @@ class Memo[T] private(val m: mutable.HashMap[RefEquality[BaseIR], T]) {
     this
   }
 
+  def bindIf(test: Boolean, ir: BaseIR, t: T): Memo[T] =
```

Comment:
This is certainly fine in the unit case, but we might not want to use this naively for every analysis if constructing `t` is not cheap (as it is for Unit), since we'll need to construct `t` even if we don't end up binding it.

An alternative is changing the signature to `def bindIf(test: Boolean, ir: BaseIR, t: => T)`, but this has other performance consequences -- Scala desugars this to a `() => T` and now a closure is allocated for every call to the function.

I think this is fine, we just won't be able to use it everywhere.

Reply:
You're right, call by name makes much more sense here. Use it where appropriate I guess. Do you know if visualvm works with scala code?

Reply:
BTW, really appreciate the explanations in comments. Thank you for taking the time.

Reply:
we typically use either YourKit or async-profiler. async-profiler is free and lighter-weight, but the last time we looked a year or two ago it didn't provide timings at the level of line symbols within functions, and as such wasn't quite as useful for profiling our enormous functions inside generated code.